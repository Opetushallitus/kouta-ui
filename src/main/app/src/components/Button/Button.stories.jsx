import React from 'react';
import { storiesOf } from '@storybook/react';
import { select, boolean } from '@storybook/addon-knobs';

import Button from './index';

storiesOf('Button', module).add('Basic', () => {
  const variant = select(
    'Variant',
    {
      Contained: 'contained',
      Outlined: 'outlined',
      Text: 'text',
    },
    'contained'
  );

  const color = select(
    'Color',
    {
      Primary: 'primary',
      Secondary: 'secondary',
      Success: 'success',
      Danger: 'danger',
    },
    'primary'
  );

  const size = select(
    'Size',
    {
      Small: 'small',
      Medium: 'medium',
    },
    'medium'
  );

  const disabled = boolean('Disabled', false);

  return (
    <Button variant={variant} color={color} size={size} disabled={disabled}>
      Button
    </Button>
  );
});
