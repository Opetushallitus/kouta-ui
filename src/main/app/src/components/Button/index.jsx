import React from 'react';
import { darken } from 'polished';
import styled, { css } from 'styled-components';

import { getThemeProp } from '../../theme';

const getOutlinedColorCss = ({ color, theme }) => {
  let outlineColor = theme.palette.primary.main;

  if (color === 'secondary') {
    outlineColor = theme.palette.secondary.main;
  }

  return `
    border-color: ${outlineColor};
    color: ${outlineColor};

    &:hover {
      border-color: ${darken(0.05, outlineColor)};
      color: ${darken(0.05, outlineColor)};
    }

    &:active {
      border-color: ${darken(0.1, outlineColor)};
      color: ${darken(0.1, outlineColor)};
    }
  `;
};

const getContainedColorCss = ({ color, theme }) => {
  let fontColor = theme.palette.primary.contrastText;
  let containColor = theme.palette.primary.main;

  if (color === 'secondary') {
    fontColor = theme.palette.secondary.contrastText;
    containColor = theme.palette.secondary.main;
  }

  return `
    border-color: ${containColor};
    background-color: ${containColor};
    color: ${fontColor};

    &:hover {
      border-color: ${darken(0.05, containColor)};
      background-color: ${darken(0.05, containColor)};
    }

    &:active {
      border-color: ${darken(0.1, containColor)};
      background-color: ${darken(0.1, containColor)};
    }
  `;
};

const getVariantCss = ({ variant }) => {
  if (variant === 'outlined') {
    return css`
      transition: color 0.25s, border-color 0.25s;
      background-color: transparent;
      ${getOutlinedColorCss}
    `;
  } else if (variant === 'contained') {
    return css`
      transition: background-color 0.25s, border-color 0.25s;
      ${getContainedColorCss}
    `;
  }
};

const ButtonBase = styled.button`
  outline: none;
  padding: 6px 16px;
  border-radius: ${getThemeProp('shape.borderRadius')};
  font-family: ${getThemeProp('typography.fontFamily')};
  font-size: 1rem;
  border: 2px solid;
  line-height: 1.5;
  white-space: nowrap;

  ${getVariantCss}

  ${({ disabled }) =>
    disabled &&
    css`
      opacity: 0.5;
    `}
`;

const Button = ({ variant = 'contained', color = 'primary', ...props }) => (
  <ButtonBase variant={variant} color={color} ref={props.innerRef} {...props} />
);

export default Button;
