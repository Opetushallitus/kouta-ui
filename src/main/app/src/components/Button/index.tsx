import React from 'react';
import VUCButton from '@opetushallitus/virkailija-ui-components/Button';

export const Button = props => {
  return (
    <div className="ButtonWrapper">
      <VUCButton {...props} />
    </div>
  );
};

export default Button;
