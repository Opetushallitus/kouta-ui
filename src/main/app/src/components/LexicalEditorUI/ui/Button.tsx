/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { ReactNode } from 'react';

import styled from 'styled-components';

import joinClasses from '../utils/joinClasses';

export const Button = styled(PlainButton)`
  .Button__root {
    padding-top: 10px;
    padding-bottom: 10px;
    padding-left: 15px;
    padding-right: 15px;
    border: 0px;
    background-color: #eee;
    border-radius: 5px;
    cursor: pointer;
    font-size: 14px;
  }
  .Button__root:hover {
    background-color: #ddd;
  }
  .Button__small {
    padding-top: 5px;
    padding-bottom: 5px;
    padding-left: 10px;
    padding-right: 10px;
    font-size: 13px;
  }
  .Button__disabled {
    cursor: not-allowed;
  }
  .Button__disabled:hover {
    background-color: #eee;
  }
`;

function PlainButton({
  'data-test-id': dataTestId,
  children,
  className,
  onClick,
  disabled,
  small,
  title,
}: {
  'data-test-id'?: string;
  children: ReactNode;
  className?: string;
  disabled?: boolean;
  onClick: () => void;
  small?: boolean;
  title?: string;
}): JSX.Element {
  return (
    <button
      disabled={disabled}
      className={joinClasses(
        'Button__root',
        disabled && 'Button__disabled',
        small && 'Button__small',
        className
      )}
      onClick={onClick}
      title={title}
      aria-label={title}
      {...(dataTestId && { 'data-test-id': dataTestId })}
    >
      {children}
    </button>
  );
}

export default Button;
