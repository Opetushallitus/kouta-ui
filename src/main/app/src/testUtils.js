import React from 'react';
import { shallow, mount } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import defaultTheme from './theme';

const makeFnWithTheme = fn => tree => {
  return fn(<ThemeProvider theme={defaultTheme}>{tree}</ThemeProvider>);
};

export const shallowWithTheme = makeFnWithTheme(shallow);

export const mountWithTheme = makeFnWithTheme(mount);

export const setMockDate = date => {
  const _Date = Date;

  global.Date.now = () => +date;

  return () => {
    global.Date = _Date;
  };
};
