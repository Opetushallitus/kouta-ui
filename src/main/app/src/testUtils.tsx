import React, { PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { ThemeProvider } from 'styled-components';

import { createRootReducer } from './state/rootReducer';
import defaultTheme from './theme';

const makeFnWithTheme = fn => tree => {
  return fn(<ThemeProvider theme={defaultTheme}>{tree}</ThemeProvider>);
};

export const renderWithTheme = makeFnWithTheme(render);

export const mountWithTheme = tree => {
  const { container } = makeFnWithTheme(render)(tree);
  return container.firstChild;
};

export const setMockDate = date => {
  const _Date = Date;

  global.Date.now = () => Number(date);

  return () => {
    global.Date = _Date;
  };
};

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = configureStore({ reducer: createRootReducer({}), preloadedState }),
    ...renderOptions
  }
) {
  // eslint-disable-next-line @typescript-eslint/ban-types
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return (
      <Provider store={store}>
        <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>
      </Provider>
    );
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from '@testing-library/react';
