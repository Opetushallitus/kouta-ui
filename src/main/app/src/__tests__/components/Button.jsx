import React from 'react';
import { mount } from 'enzyme';

import Button from '../../components/Button';
import { mountWithTheme } from '../../testUtils';

test('renders correctly with variant', () => {
  expect(
    mountWithTheme(<Button variant="contained">Contained button</Button>),
  ).toMatchSnapshot();

  expect(
    mountWithTheme(<Button variant="outlined">Outlined button</Button>),
  ).toMatchSnapshot();
});

test('renders correctly with color', () => {
  expect(
    mountWithTheme(<Button color="primary">Primary button</Button>),
  ).toMatchSnapshot();

  expect(
    mountWithTheme(<Button color="secondary">Secondary button</Button>),
  ).toMatchSnapshot();
});
