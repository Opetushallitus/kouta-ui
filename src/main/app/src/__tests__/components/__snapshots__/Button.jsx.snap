// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with color 1`] = `
.c0 {
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: sans-serif;
  font-size: 1rem;
  border: 2px solid;
  line-height: 1.5;
  -webkit-transition: background-color 0.25s,border-color 0.25s;
  transition: background-color 0.25s,border-color 0.25s;
  border-color: #2da0c7;
  background-color: #2da0c7;
  color: #ffffff;
}

.c0:hover {
  border-color: #288fb2;
  background-color: #288fb2;
}

.c0:active {
  border-color: #247f9d;
  background-color: #247f9d;
}

<styled.button
  color="primary"
  variant="contained"
>
  <button
    className="c0"
    color="primary"
  >
    Primary button
  </button>
</styled.button>
`;

exports[`renders correctly with color 2`] = `
.c0 {
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: sans-serif;
  font-size: 1rem;
  border: 2px solid;
  line-height: 1.5;
  -webkit-transition: background-color 0.25s,border-color 0.25s;
  transition: background-color 0.25s,border-color 0.25s;
  border-color: #999999;
  background-color: #999999;
  color: #ffffff;
}

.c0:hover {
  border-color: #8c8c8c;
  background-color: #8c8c8c;
}

.c0:active {
  border-color: #808080;
  background-color: #808080;
}

<styled.button
  color="secondary"
  variant="contained"
>
  <button
    className="c0"
    color="secondary"
  >
    Secondary button
  </button>
</styled.button>
`;

exports[`renders correctly with variant 1`] = `
.c0 {
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: sans-serif;
  font-size: 1rem;
  border: 2px solid;
  line-height: 1.5;
  -webkit-transition: background-color 0.25s,border-color 0.25s;
  transition: background-color 0.25s,border-color 0.25s;
  border-color: #2da0c7;
  background-color: #2da0c7;
  color: #ffffff;
}

.c0:hover {
  border-color: #288fb2;
  background-color: #288fb2;
}

.c0:active {
  border-color: #247f9d;
  background-color: #247f9d;
}

<styled.button
  color="primary"
  variant="contained"
>
  <button
    className="c0"
    color="primary"
  >
    Contained button
  </button>
</styled.button>
`;

exports[`renders correctly with variant 2`] = `
.c0 {
  outline: none;
  padding: 6px 16px;
  border-radius: 4px;
  font-family: sans-serif;
  font-size: 1rem;
  border: 2px solid;
  line-height: 1.5;
  -webkit-transition: color 0.25s,border-color 0.25s;
  transition: color 0.25s,border-color 0.25s;
  background-color: transparent;
  border-color: #2da0c7;
  color: #2da0c7;
}

.c0:hover {
  border-color: #288fb2;
  color: #288fb2;
}

.c0:active {
  border-color: #247f9d;
  color: #247f9d;
}

<styled.button
  color="primary"
  variant="outlined"
>
  <button
    className="c0"
    color="primary"
  >
    Outlined button
  </button>
</styled.button>
`;
